### PR TITLE
Added trigger functionality to Dolt

### PR DESCRIPTION
**Pull request by Hydrocharged**

### CODE REVIEW

**Comment by andrew-wm-arthur** [comment]
What are the tradeoffs between adding triggers to the `dolt_schemas` table vs creating a separate system table for them?
I'm concerned about the compatibility between existing `dolt_schemas` tables and new versions with the updated schema. 
Also we need tests for diffing/merging triggers. Specifically, the cycle detection we do on `CreateTrigger()` should throw a conflict when we merge triggers across branches.

File: go/libraries/doltcore/sqle/schema_table.go [hunk 1]
```diff
@@ -15,65 +15,174 @@
...
+
 	"github.com/dolthub/go-mysql-server/sql"
 
 	"github.com/dolthub/dolt/go/libraries/doltcore/doltdb"
+	"github.com/dolthub/dolt/go/libraries/doltcore/row"
+	"github.com/dolthub/dolt/go/store/types"
 )
 
 // The fixed schema for the `dolt_schemas` table.
 func SchemasTableSchema() sql.Schema {
```
Comment (andrew-wm-arthur) [inline go/libraries/doltcore/sqle/schema_table.go:29]: We need to generate this schema as a Dolt `schema.Schema` in order to hard-code its tags. Currently it's are generated via the deterministic tag method, which is not stable when modifying the schema. See the [query catalog table](https://github.com/dolthub/dolt/blob/4f20d5f78c80885bc0a26a5a51d85cee6b5c14d7/go/libraries/doltcore/sqle/query_catalog_table.go#L29) for an example.

File: go/libraries/doltcore/doltdb/system_table.go [hunk 2]
```diff
@@ -204,15 +204,17 @@ const (
...
 	SchemasTablesTypeCol = "type"
-
-	// // The name of the database entity.
+	// The name of the database entity.
 	SchemasTablesNameCol = "name"
 	// The schema fragment associated with the database entity.
 	// For example, the SELECT statement for a CREATE VIEW.
 	SchemasTablesFragmentCol = "fragment"
+	// The name of the index that is on the table.
+	SchemasTablesIndexName = "fragment_name"
```
Comment (andrew-wm-arthur) [inline go/libraries/doltcore/doltdb/system_table.go:217]: We also need to update the next `const` block with new tags. However, the current tags in that block are currently incorrect/unused. See the comment in `sqle/schema_table.go` for details.

**Review by andrew-wm-arthur — COMMENTED** [review]
Per our conversation, there are some compatibility concerns stemming from the existing `dolt_schemas` table schema implementation. 
Also, we need some merge/diff tests. Trigger ordering and ref-cycle detection could be tricky in the merge case.

File: go/libraries/doltcore/sqle/schema_table.go [hunk 3]
```diff
@@ -15,65 +15,174 @@
...
+	tbl, found, err := db.GetTableInsensitiveWithRoot(ctx, root, doltdb.SchemasTableName)
 	if err != nil {
 		return nil, err
 	}
+	var rowsToAdd []sql.Row
 	if found {
-		return tbl.(*WritableDoltTable), nil
+		schemasTable := tbl.(*WritableDoltTable)
+		// Old schemas table does not contain the `id` column.
+		if !tbl.Schema().Contains(doltdb.SchemasTablesIdCol, doltdb.SchemasTableName) {
```
Comment (zachmu) [inline go/libraries/doltcore/sqle/schema_table.go:52]: Suggest extracting a method for this migration logic

**Review by zachmu — APPROVED** [review]
LG, just add the logic about using predefined tags in the system space for the schemas table like we talked offline, and this is good to go

**Comment by Hydrocharged** [comment]
@andrew-wm-arthur Since this all resides in the `dolt_schemas` table anyway, I actually don't think we need to worry about catching ordering and cycle detection issues. It's completely valid to write directly into the `dolt_schemas` table, including writing a trigger that would cause those issues, and thus it can be fixed just the same.

On another note, while writing a merge test, I found an unrelated issue with how we handle conflicts. Rather than fixing that and possibly delaying getting triggers in, I'll just merge this and fix that in another PR.

**Pull request merged**